File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;

class Solution {
    public String[] solution(String[] record) {
        String ENTER_MESSAGE = "님이 들어왔습니다.";
        String LEAVE_MESSAGE = "님이 나갔습니다.";
        
        HashMap<String, String> uidNicknameMap = new HashMap<>();
        List<String[]> finalMessage = new ArrayList<>();
        
        for(String rec: record){
            String[] ipt = rec.split(" ");
            String command = ipt[0];
            String uid = ipt[1];
            if(command.equals("Enter")){
                String nickname = ipt[2];                
                uidNicknameMap.put(uid, nickname);
                finalMessage.add(new String[] {"Enter", uid});
            } else if(command.equals("Change")){
                String nickname = ipt[2];                
                uidNicknameMap.put(uid, nickname);
            } else {
                finalMessage.add(new String[] {"Leave", uid});
            }
        }
        
        
        String[] answer = new String[finalMessage.size()];
        
        for(int i = 0; i < answer.length; i++){
            String uid = finalMessage.get(i)[1];
            if(finalMessage.get(i)[0].equals("Enter")){
                answer[i] = uidNicknameMap.get(uid) + ENTER_MESSAGE;
            } else {
                answer[i] = uidNicknameMap.get(uid) + LEAVE_MESSAGE;
            }
            
        }
        return answer;
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    answer = []
    uid_nickname_map = {}
    
    for rec in record:
        cur_input = list(rec.split(" "))
        if len(cur_input) > 2:
            command, uid, nickname = cur_input
        else:
            command, uid  = cur_input
        
        if command == "Enter":
            uid_nickname_map[uid] = nickname
            answer.append([uid, "Enter"])
        elif command == "Leave":            
            answer.append([uid, "Leave"])
        elif command == "Change":
            uid_nickname_map[uid] = nickname
    
    for i in range(len(answer)):
        uid, command = answer[i]
        if command == "Enter":
            answer[i] = get_enter_message(uid_nickname_map[uid])
        elif command == "Leave":
            answer[i] = get_leave_message(uid_nickname_map[uid])
    
    return answer


def get_enter_message(nickname: str):
    return f"{nickname}님이 들어왔습니다."


def get_leave_message(nickname: str):
    return f"{nickname}님이 나갔습니다."

